modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/CurrencySnapshotCreator.scala

```diff
@@ -50,137 +54,181 @@ trait CurrencySnapshotCreator[F[_]] {
-        }
+          currentEpochProgress = trigger match {
+            case EventTrigger => lastArtifact.epochProgress
+            case TimeTrigger  => lastArtifact.epochProgress.next
+          }
+          lastActiveTips <- lastArtifact.activeTips
+          lastDeprecatedTips = lastArtifact.tips.deprecated
 
-        (deprecated, remainedActive, accepted) = getUpdatedTips(
-          lastActiveTips,
-          lastDeprecatedTips,
-          acceptanceResult,
-          currentOrdinal
-        )
+          maybeLastDataApplication = lastArtifact.data
 
-        (height, subHeight) <- getHeightAndSubHeight(lastArtifact, deprecated, remainedActive, accepted)
-
-        artifact = CurrencyIncrementalSnapshot(
-          currentOrdinal,
-          height,
-          subHeight,
-          lastArtifactHash,
-          accepted,
-          acceptedRewardTxs,
-          SnapshotTips(
-            deprecated = deprecated,
-            remainedActive = remainedActive
-          ),
-          stateProof,
-          currentEpochProgress,
-          maybeNewDataState
-        )
-      } yield
-        CurrencySnapshotCreationResult(
-          artifact,
-          CurrencySnapshotContext(lastContext.address, snapshotInfo),
-          awaitingBlocks.toSet,
-          rejectedBlocks.toSet
-        )
+          maybeLastDataState <- (maybeLastDataApplication, maybeDataApplicationService).mapN {
+            case (lastDataApplication, (_, service)) =>
+              service
+                .deserializeState(lastDataApplication)
+                .flatTap {
+                  case Left(err) => logger.warn(err)(s"Cannot deserialize custom state")
+                  case Right(a)  => logger.info(s"Deserialized state: $a")
+                }
+                .map(_.toOption)
+                .handleErrorWith(err =>
+                  logger.error(err)(s"Unhandled exception during deserialization data application, fallback to empty state") >>
+                    none[DataState].pure[F]
+                )
+          }.flatSequence
+
+          maybeNewDataState <- (maybeDataApplicationService, maybeLastDataState).mapN {
+            case (((nodeContext, service), lastState)) =>
+              implicit val context: L0NodeContext[F] = nodeContext
+              NonEmptyList
+                .fromList(dataBlocks.distinctBy(_.value.roundId))
+                .map(_.flatMap(_.value.updates))
+                .map { updates =>
+                  service
+                    .validateData(lastState, updates)
+                    .flatTap { validated =>
+                      Applicative[F].unlessA(validated.isValid)(logger.warn(s"Data application is invalid, errors: ${validated.toString}"))
+                    }
+                    .map(_.isValid)
+                    .handleErrorWith(err =>
+                      logger.error(err)(s"Unhandled exception during validating data application, assumed as invalid") >> false.pure[F]
```

Comment:
```suggestion
                      logger.error(err)("Unhandled exception during validating data application, assumed as invalid").as(false)
```
Also, should this also be a WARN instead of ERROR?

Reply:
@marcinwadon - it is your part. Do you agree?